lib/Sema/CSSimplify.cpp

```diff
@@ -3626,18 +3626,34 @@ ConstraintSystem::matchExistentialTypes(Type type1, Type type2,
 
           if (last.is<LocatorPathElt::ApplyArgToParam>()) {
             auto proto = protoDecl->getDeclaredInterfaceType();
-            auto *fix = AllowArgumentMismatch::create(
-                *this, type1, proto, getConstraintLocator(anchor, path));
-
+            ConstraintFix *fix;
             // Impact is 2 here because there are two failures
             // 1 - missing conformance and 2 - incorrect argument type.
             //
             // This would make sure that arguments with incorrect
             // conformances are not prioritized over general argument
             // mismatches.
-            if (recordFix(fix, /*impact=*/2))
+            unsigned impact;
+            if (type1->isOptional()) {
+              auto unwrappedType = type1->lookThroughAllOptionalTypes();
+              auto matchTypeResult =
+                  matchTypes(unwrappedType, type2, ConstraintKind::Conversion,
+                             subflags, locator);
+              if (matchTypeResult.isSuccess()) {
+                // Impact here is a 1 because there is only one failure before
+                // we need to force type1 to be
+                impact = 1;
+                fix = ForceOptional::create(*this, type1, proto,
+                                            getConstraintLocator(locator));
```

Comment:
To simplify this code you can `recordFix` directly here with return and add a `break` right after, that way you won't need to use `ConstraintFix *` and impact variables.

Reply:
With this approach - I am seeing both errors returned --> 

/Users/dylanperry/dev/ios/open-source/Swift/Test Files/Test.swift:8:10: error: argument type 'any P' does not conform to expected type 'P'
  takesP(value) // error: Argument type '(any P)?' does not conform to expected type 'P'
         ^
               as P
/Users/dylanperry/dev/ios/open-source/Swift/Test Files/Test.swift:8:10: error: value of optional type '(any P)?' must be unwrapped to a value of type 'any P'
  takesP(value) // error: Argument type '(any P)?' does not conform to expected type 'P'
         ^
/Users/dylanperry/dev/ios/open-source/Swift/Test Files/Test.swift:8:10: note: coalesce using '??' to provide a default when the optional value contains 'nil'
  takesP(value) // error: Argument type '(any P)?' does not conform to expected type 'P'
         ^
               ?? <#default value#>
/Users/dylanperry/dev/ios/open-source/Swift/Test Files/Test.swift:8:10: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
  takesP(value) // error: Argument type '(any P)?' does not conform to expected type 'P'
         ^
              !

Reply:
Added a comment about that bellow. This is cause the way we are calling `simplifyConformsToConstraint` we should pass the type and protocol declaration. See the bellow comment =]

Reply:
@LucianoPAlmeida  Thank you! I am still learning -- appreciate the patience :)